File: typescript/decorator/tesla.ts
abstract class Car {
	public description: string;
	public getDescription(): string {
		return this.description;
	}
	public abstract cost(): number;
}

class ModelS extends Car {
	public description: string = "This is my Model S Tesla";
	public getDescription(): string {
		return this.description;
	}
	public cost(): number {
		return 70_000;
	}
}

class ModelX extends Car {
	public description: string = "This is my Model X Tesla";
	public getDescription(): string {
		return this.description;
	}
	public cost(): number {
		return 75_000;
	}
}

abstract class OptionalCarFeature extends Car {
	public car: Car;
	public abstract getDescription(): string;
	public abstract cost(): number;
}

class EnhancedAutoPilot extends OptionalCarFeature {
	public car: Car;
	constructor(car: Car) {
		super();
		this.car = car;
	}
	public getDescription(): string {
		return this.car.getDescription();
	}
	public cost(): number {
		return this.car.cost() + 5000;
	}
}

class RearFacingSeats extends OptionalCarFeature {
	public car: Car;
	constructor(car: Car) {
		super();
		this.car = car;
	}
	public getDescription(): string {
		return this.car.getDescription();
	}
	public cost(): number {
		return this.car.cost() + 2500;
	}
}

let myTeslaS = new ModelS();
myTeslaS = new EnhancedAutoPilot(myTeslaS);
console.log("myTeslaS", myTeslaS.getDescription(), myTeslaS.cost());
let myTeslaX = new ModelX();
myTeslaX = new RearFacingSeats(myTeslaX);
console.log("myTeslaX", myTeslaX.getDescription(), myTeslaX.cost());

File: python/decorator/tesla.py
from abc import ABC, abstractclassmethod
class Car(ABC):
    @abstractclassmethod
    def getDescription():
        pass
    @abstractclassmethod
    def cost():
        pass

class ModelS(Car):
    def __init__(self):
        self.description = "This is my Model S Tesla"

    def getDescription(self):
        return self.description

    def cost(self):
        return 70_000

class ModelX(Car):
    def __init__(self):
        self.description = "This is my Model X Tesla"

    def getDescription(self):
        return self.description

    def cost(self):
        return 75_000

class OptionalCarFeature(Car):
    def getDescription(self):
        pass
    
    def cost(self):
        pass
    
class EnhancedAutoPilot(OptionalCarFeature):
    def __init__(self, car):
        self.car = car
    
    def getDescription(self):
        return self.car.getDescription()
    
    def cost(self):
        return self.car.cost() + 2500
    
class RearFacingSeats(OptionalCarFeature):
    def __init__(self, car):
        self.car = car
    
    def getDescription(self):
        return self.car.getDescription()
    
    def cost(self):
        return self.car.cost() + 5000
    
myModelS = ModelS()
myModelS = EnhancedAutoPilot(myModelS)
print(myModelS.getDescription(), myModelS.cost())
myModelX = ModelX()
myModelX = RearFacingSeats(myModelX)
print(myModelX.getDescription(), myModelX.cost())